sihl/src/index.mld

```diff
@@ -731,6 +731,77 @@ If you have complex pre-conditions, you should move the service calls to the [da
 
 {1 Digging Deeper}
 
+{2 Pagination & Search}
+
+Displaying things and collections of things is a common thing in web development. Often we don't want to display all the items that are stored, but present the user a nice interface to search collections.
+
+Sihl services usually follow a convention where they expose a function
+
+{[
+val search:
+  ?sort:[ `Asc | `Desc ] ->
+  ?filter:string ->
+  ?limit:int ->
+  ?offset:int ->
+  unit ->
+  (t list * int) Lwt.t
+]}
+
+that can be used to fetch, sort and filter a partial view on the whole collection of [t].
+
+To implement your own function with this signature, use {!Sihl.Database.prepare_search_request} and {!Sihl.Database.run_search_request}:
+
+{[
+  let filter_fragment =
+    {sql|
+        WHERE user_users.email LIKE $1
+          OR user_users.username LIKE $1
+          OR user_users.status LIKE $1 |sql}
+  ;;
+
+  let search_query =
+    {sql|
+        SELECT
+          uuid,
+          email,
+          username,
+          password,
+          status,
+          admin,
+          confirmed,
+          created_at,
+          updated_at
+        FROM user_users |sql}
+  ;;
+
+  (* We need to escape this SQL query because
+     it breaks syntax highlghting of the documentation *)
+  let count_query = {sql| SELECT COUNT(\*\) FROM user_users |sql}
+
+  let request =
+    Sihl.Database.prepare_search_request
+      ~search_query
+      ~count_query
+      ~filter_fragment
+      ~sort_by_field:"id"
+      user
+  ;;
+
+  let search sort filter ~limit ~offset =
+    Sihl.Database.query (fun connection ->
+        let module Connection = (val connection : Caqti_lwt.CONNECTION) in
+        Sihl.Database.run_search_request
+          connection
+          request
+          sort
+          filter
+          ~limit
+          ~offset)
```

Comment:
Maybe mention somewhere that `offset` is slow?

Reply:
Added a note.

Reply:
Thanks!